gobblin-core/src/main/java/gobblin/converter/csv/CsvToJsonConverterV2.java

```diff
@@ -0,0 +1,169 @@
+   *   "Street number", "City" , "Country", "Zip code"
+   *
+   * - Custom order
+   *   0,1,3
+   * - Output JSON (Key is derived from input schema)
+   *   {"street_number" : "Street number", "city" : "City" , "zip_code" : "Zip code" }
+   *
+   * - Custom order (adding null when negative index is defined)
+   *   0,1,-1,3
+   * - Output JSON
+   *   {"street_number" : "Street number", "city" : "City" , "Country" : null, "zip_code" : "Zip code" }
+   *
+   *
+   * {@inheritDoc}
+   * @see gobblin.converter.Converter#convertRecord(java.lang.Object, java.lang.Object, gobblin.configuration.WorkUnitState)
+   */
+  @Override
+  public Iterable<JsonObject> convertRecord(JsonArray outputSchema, String[] inputRecord, WorkUnitState workUnit)
+      throws DataConversionException {
+
+    JsonObject outputRecord = null;
+    if (!customOrder.isEmpty()) {
+      outputRecord = createOutput(outputSchema, inputRecord, customOrder);
+    } else {
+      outputRecord = createOutput(outputSchema, inputRecord);
+    }
+
+    if (LOG.isDebugEnabled()) {
+      LOG.debug("Converted into " + outputRecord);
+    }
+    return new SingleRecordIterable<JsonObject>(outputRecord);
+  }
+
+  @VisibleForTesting
+  JsonObject createOutput(JsonArray outputSchema, String[] inputRecord) {
+    Preconditions.checkArgument(outputSchema.size() == inputRecord.length, "# of columns mismatch. Input "
+        + inputRecord.length + " , output: " + outputSchema.size());
+    JsonObject outputRecord = new JsonObject();
+
+    for (int i = 0; i < outputSchema.size(); i++) {
+      String key = outputSchema.get(i).getAsJsonObject().get(COLUMN_NAME_KEY).getAsString();
+
+      if (StringUtils.isEmpty(inputRecord[i]) || JSON_NULL_VAL.equalsIgnoreCase(inputRecord[i])) {
+        outputRecord.add(key, JsonNull.INSTANCE);
+      } else {
+        outputRecord.addProperty(key, inputRecord[i]);
+      }
+    }
+
+    return outputRecord;
+  }
+
+  @VisibleForTesting
+  JsonObject createOutput(JsonArray outputSchema, String[] inputRecord, List<String> customOrder) {
+
+    Preconditions.checkArgument(outputSchema.size() == customOrder.size(), "# of columns mismatch. Input "
+        + inputRecord.length + " , output: " + outputSchema.size());
+    JsonObject outputRecord = new JsonObject();
+    Iterator<JsonElement> outputSchemaIterator = outputSchema.iterator();
+    Iterator<String> customOrderIterator = customOrder.iterator();
+
+    while(outputSchemaIterator.hasNext() && customOrderIterator.hasNext()) {
+      String key = outputSchemaIterator.next().getAsJsonObject().get(COLUMN_NAME_KEY).getAsString();
+      int i = Integer.parseInt(customOrderIterator.next());
+      Preconditions.checkArgument(i < inputRecord.length, "Index out of bound detected in customer order. Index: " + i + " , # of CSV columns: " + inputRecord.length);
+      if (i < 0 || null == inputRecord[i] || JSON_NULL_VAL.equalsIgnoreCase(inputRecord[i])) {
+        outputRecord.add(key, JsonNull.INSTANCE);
```

Comment:
Since you iterate over and add in same order as outputSchema, aren't you maintaining the same order by with mismatched inputvalue rather? Lets discuss this offline, maybe I am not understanding this correctly.